flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/TieredStorageProducerClientTest.java

```diff
@@ -0,0 +1,187 @@
+
+        TieredStorageProducerClient tieredStorageProducerClient =
+                createTieredStorageProducerClient(
+                        numSubpartitions, Collections.singletonList(tierProducerAgent));
+
+        int numWriteRecords = 0;
+        for (int j = 0; j < numSubpartitions; j++) {
+            for (int i = 0; i < random.nextInt(maxNumToWriteRecordsPerSubpartition); i++) {
+                tieredStorageProducerClient.write(
+                        generateRandomData(bufferSize, random),
+                        new TieredStorageSubpartitionId(j),
+                        Buffer.DataType.DATA_BUFFER,
+                        isBroadcast);
+                numWriteRecords++;
+            }
+        }
+
+        int numExpectedBuffers = isBroadcast ? numWriteRecords * numSubpartitions : numWriteRecords;
+        assertThat(tierProducerAgent.numTotalReceivedBuffers()).isEqualTo(numExpectedBuffers);
+    }
+
+    @TestTemplate
+    void testTierCanNotStartNewSegment() throws IOException {
+        int numSubpartitions = 10;
+        int bufferSize = 1024;
+        Random random = new Random();
+
+        TestingTierProducerAgent tierProducerAgent = new TestingTierProducerAgent();
+        tierProducerAgent.setTryStartNewSegmentReturnValueSupplier(() -> false);
+        tierProducerAgent.setTryWriteReturnValueSupplier(() -> new Boolean[] {false, true});
+        TieredStorageProducerClient tieredStorageProducerClient =
+                createTieredStorageProducerClient(
+                        numSubpartitions, Collections.singletonList(tierProducerAgent));
+
+        assertThatThrownBy(
+                        () ->
+                                tieredStorageProducerClient.write(
+                                        generateRandomData(bufferSize, random),
+                                        new TieredStorageSubpartitionId(0),
+                                        Buffer.DataType.DATA_BUFFER,
+                                        isBroadcast))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessageContaining("Failed to choose a storage tier");
+
+        tierProducerAgent.setTryStartNewSegmentReturnValueSupplier(() -> true);
+        tieredStorageProducerClient.write(
+                generateRandomData(bufferSize, random),
+                new TieredStorageSubpartitionId(0),
+                Buffer.DataType.DATA_BUFFER,
+                isBroadcast);
+
+        assertThat(tierProducerAgent.numTotalReceivedBuffers())
+                .isEqualTo(isBroadcast ? numSubpartitions : 1);
+    }
+
+    @TestTemplate
+    void testClose() {
+        int numSubpartitions = 10;
+
+        TestingTierProducerAgent tierProducerAgent = new TestingTierProducerAgent();
+        tierProducerAgent.setTryStartNewSegmentReturnValueSupplier(() -> true);
+        tierProducerAgent.setTryWriteReturnValueSupplier(() -> new Boolean[] {false, true});
```

Comment:
Are these two lines necessary?  Because we haven't even call the `write` method, why do we care about them?

Reply:
Fixed.